### PR TITLE
Improve total_spanning_tree_weight

### PR DESCRIPTION
**Pull request by peijenburg**
Following #7065

Improved `total_spanning_tree_weights`:

- Added the computation for directed graphs
- Added examples
- Added references
- Added tests
- Prevented corner case (like graphs without nodes or edges or disconnected graphs)

### CODE REVIEW

**Comment by peijenburg** [comment]
@dschult Sorry for the ping, but if you have some time, a review is appreciated :)

File: networkx/linalg/laplacianmatrix.py
Comment (dschult) [inline networkx/linalg/laplacianmatrix.py:228]: `if not graph_is_directed` rather than `== False`.

File: networkx/linalg/laplacianmatrix.py
Comment (dschult) [inline networkx/linalg/laplacianmatrix.py:255]: Can you add `axis=` ?  Also a real nit-picky thing: switch the order of the if/else clause defining `i` so that you can use the `if graph_is_directed:` without any negation. Easier to read... ?maybe?

File: networkx/linalg/tests/test_laplacian.py
Comment (dschult) [inline networkx/linalg/tests/test_laplacian.py:272]: Why are negative weights ignored instead of an exception raised. Doesn't that cause surprises? It might be better to raise an exception so they know there is a problem and can fix it. Otherwise they may never know they had a negative weight.

**Review by dschult — COMMENTED** [review]
Thanks for giving this function a sprucing up!
We will be renaming it soon too. But this makes it more presentable. It likely didn't have much exposure before. :(

Most of these comments are nit-picky choice of wording or syntax. Feel free to push back or ask why.
:)

File: networkx/linalg/laplacianmatrix.py
Comment (peijenburg) [reply networkx/linalg/laplacianmatrix.py:158]: I have reformulated the part about directed graphs with as less jargon as possible, but if there's still anything unclear, I'm open to better suggestions.

**Review by dschult — COMMENTED** [review]
I'm still having trouble reading this code easily. Could you simplify the test code (one example of what I mean is below)?  And for the code in the function, it seems like every single section has an `if` clause to check for directed. The previous code was about 3 lines long so this simplicity should be reflected in the new solution somehow.

Can we try one if clause:
```python
if not G.is_directed():
    <do the 3 line undirected code, including returning the value>
else:
    <do all the directed stuff and return the value>
```
Even the checks for connectedness can be done this way. 

My second question is why this PR changes the behavior for disconnected graphs.  The result is correct:  The number of spanning trees is zero (a tree can't cover all the nodes) and the determinant of the laplacian is also zero. 

Thanks!!

File: networkx/linalg/laplacianmatrix.py
Comment (peijenburg) [reply networkx/linalg/laplacianmatrix.py:260]: The original implementation did not allow for weighted graphs, thus the absolute value was allowed. 
For graphs with negative weights, the number of spanning trees can be negative, see test `test_tstw_negative_weight`

**Comment by peijenburg** [comment]
Changes

- Simplified tests
- I do not particularly agree that the code will be easier to read if I implement your suggestion. I tried to separate the code into three clear steps:
-- Step 1: compute the (directed) weighted Laplacian matrix
-- Step 2: remove one row/column
-- Step 3: compute the determinant
To me this seems very clear and intuitive, but if you have any suggestions, please let me know.
- The number of spanning trees in disconnected graphs should indeed return 0, my apologies. I modified the code such that disconnected graphs return 0.

File: networkx/linalg/laplacianmatrix.py
Comment (dschult) [reply networkx/linalg/laplacianmatrix.py:260]: One of the important properties of the laplacian matrix is that it is positive semi-definite. When you consider negative weights the laplacian matrix is no longer positive semi-definite.  
xLx^T  = sum_{(i,j)\in E} w_{ij} * (x_i-x_j)^2  and this is positive for positive weights (and real vectors x).

The reason we use the `abs` is to remove the possibility of a negative result due to round-off errors. If there are such negative results due to round-off, they will be small so we reasoned that taking the `abs` will introduce an error of the size of round-off error. And the result will be positive.

BUT...  I don't know if the Kirchhoff theorem relies on the positive semi-definite feature of Laplacians. I know that there can't be a negative number of spanning trees. :)  But your test shows that with negative weights there can be negative cofactors of the laplacian. I guess the question is whether the theorem depends on the weights being positive or not. Do you have some insight into that question?

Thanks...

Comment (peijenburg) [reply networkx/linalg/laplacianmatrix.py:260]: @dschult I understand the reasoning for using `abs`, but it is generally not correct. Negative weights are allowed, because the theorem does not make use of the positive definiteness of the Laplacian matrix. Of course, negative weights are not commonly used, but the generalisation is straightforward and I couldn't find any counter-examples where the number of spanning trees suddenly becomes negative for non-negative weighted graphs (due to rounding errors) so I thought this was a good solution. 

Happy to hear your take on this.

**Comment by dschult** [comment]
Let's leave off the absolute value -- round-off error is something we have to live with for floating point computations. And who knows if someone might actually need to compute this for weights that give a negative result.  And it does make sense that the theorem's formula holds whether the matrix is positive definite or not, it's just not as clear what the value of the result is. Thanks!

**Comment by peijenburg** [comment]
@dschult Thanks! So I guess this PR is ready for merging?

**Comment by dschult** [comment]
I'm working on yet another review... :}

**Review by dschult — COMMENTED** [review]
OK... I've got a "shell function" framework set up.  I decided it would be easier to provide my review suggestions via a commit. So please take a look and change anything you don't want in here. You will have to pull the commit to your local machine before making more commits to push back up.

Question:  I think the `root` node has to be able to reach (or maybe it is be reached by) every node in the graph. Should we be checking that it is a valid root? If we forced strong_connected, that wouldn't be an issue. Right? but weakly connected allows some nodes which cannot be a root. How should we handle this?  Perhaps instead of checking weakly connected we can see if the number of descendants is the same and the number of nodes. But maybe I've mixed something up too... Thoughts?

I've written the new function to get rid of the if-directed statements and to avoid scipy.sparse since we convert to numpy array anyway. The old function (total_...) still has the other code if you want to compare side-by-side. 

I tried to make the docs more complete. Let me know any corrections or preference here --or just make them yourself.

The doc_string says we raise a NetworkXPointlessConcept, so I made that change and had to change a test or two. This meant changing a test (for not root present) to that it would catch that instead of a null graph.  There is one test that fails now. It has a root that is not a valid root, but I don't understand how it passed before.

**Comment by peijenburg** [comment]
@dschult 

Amazing, thanks a lot for your help! Your code looks great. I've made a few changes, feel free to revert them if necessary.

> Question: I think the root node has to be able to reach (or maybe it is be reached by) every node in the graph. Should we be checking that it is a valid root? If we forced strong_connected, that wouldn't be an issue. Right? but weakly connected allows some nodes which cannot be a root. How should we handle this? Perhaps instead of checking weakly connected we can see if the number of descendants is the same and the number of nodes. But maybe I've mixed something up too... Thoughts?

We could check explicitly that every node can reach node `root`, but I don't think that's required. Have a look at the test `test_nst_directed_weighted`.
And you are right, it is required that there is a path from each node to the root node. I have adjusted the documentation accordingly, but wasn't sure about the antonym for decendants. I used `successors`, but there's a probably better graph-related word for that?

> I've written the new function to get rid of the if-directed statements and to avoid scipy.sparse since we convert to numpy array anyway. The old function (total_...) still has the other code if you want to compare side-by-side.

Fine with me. Just a thought: For directed graphs, we compute G_laplacian = A - D where A is sparse and D is not. Is this a problem?

> I tried to make the docs more complete. Let me know any corrections or preference here --or just make them yourself.

Much better now, thanks!

>The doc_string says we raise a NetworkXPointlessConcept, so I made that change and had to change a test or two. This meant changing a test (for not root present) to that it would catch that instead of a null graph. There is one test that fails now. It has a root that is not a valid root, but I don't understand how it passed before.

Thanks! There was a problem in your code, where `G.out_degree` was missing `nodelist`, which I fixed now.

Final question: Should we remove the documentation at `total_spanning_tree_weight` and directly point to `number_of_spanning_trees` ? That sounds more logical to me, but I don't know about the usual networkx practice. Also the tests at `total_spanning_tree_weight` can be removed IMO.

**Review by dschult — COMMENTED** [review]
Thanks for rewriting and catching tricky bugs, typos and spelling errors! Very nice. :}

The antonym for descendant is ancestor (at least in NetworkX) and we use successor/predecessor to mean direct descendant and direct ancestor. Also important, the terms `root` and `leaf` in NX go up the tree. That is, a root only has outgoing edges and a leaf only has incoming edges. I think that is the opposite of this algorithm/paper. See my comment below.

Having A be sparse and D dense is OK for subtraction. The result is a numpy array (dense).  So, if we want the laplacian to be sparse, we should make the diagonal matrix sparse also. In our case, we use numpy to compute the determinant so it all has to be converted to numpy anyway.  Scipy has a determinant function for numpy arrays too, but the docs suggest that it uses the same methods (LAPACK code getrf). Either way we need dense.  Numerical error computing the determinant with large networks is a well-known problem so numpy (and others) provide functions to return the log of the determinant. I think we will have those kind of errors for large networks with this code, but that is the topic for a future separate PR.

I think we should leave the doc_strings and tests in place for `total_spanning_tree_weight`. We will remove them when the function is removed. 
BUT: We **should** put the deprecation warning text into the doc_string and we should add a test that checks for that warning. That also involves flagging the tests in `conftest.py` so they don't fail due to the deprecation warning. The instructions are at [contributing workflow](https://networkx.org/documentation/stable/developer/contribute.html#development-workflow) number 6. But I can do it if you prefer.

Thanks again for the updates!

File: networkx/algorithms/tree/mst.py
Comment (peijenburg) [reply networkx/algorithms/tree/mst.py:1205]: No. Based on your suggestion, I've changed it to return 0. That also makes more sense, because the number of spanning trees of a disconnected graph has a perfectly valid answer: 0

File: networkx/algorithms/tree/mst.py
Comment (peijenburg) [reply networkx/algorithms/tree/mst.py:1146]: I don't think the directed version is used often. So we can adhere to the usual NX nomenclature: I have changed `root` to `leaf` everywhere

**Comment by peijenburg** [comment]
@dschult Thanks for sharing your thoughts!

I've changed `root` to `leaf` everywhere, as the literature on the number of spanning trees in directed graphs isn't very large (in my limited understanding), so we can safely adhere to the standard NX guidelines for consistency.

If you could take care of the deprecation warnings, that'd be great!

**Comment by dschult** [comment]
Thanks for these changes -- I will take care of the deprecations.

I wasn't expected the name change from root to leaf. I was expecting the name root to stay (`root` is a common term for trees, with a tree that has a root called a `rooted tree`). But the description of the meaning of "root" should shift from incoming edges to outgoing edges. The arborecenses include, for each node, a unique path from the root to that node.

The code currently computes the paths following successors (out-edges) and using `out_degree` of the nodes when computing the laplacian. Doesn't that mean the root is a node from which outgoing edges exist connecting it to all the other nodes?  Let's figure out what the code computes -- then we can align the docs with it.

**Comment by peijenburg** [comment]
Thanks @dschult but perhaps I wasn't clear. Let me try to explain.

> The code currently computes the paths following successors (out-edges) and using out_degree of the nodes when computing the laplacian. Doesn't that mean the root is a node from which outgoing edges exist connecting it to all the other nodes?

No, it doesn't. Have a look at `test_nst_directed_weighted`. The graph has edges 1->2, 1->3 and 2->3. Only choosing `leaf=3` results in a non-zero number of spanning trees, indicating that the leaf node must actually be, what I would call, a `sink`, namely there exists a directed path from every node to the sink.
Any other choice for `leaf` results in zero spanning trees, as expected, because there is no path from node 3 to node 1 or node 2.

**Comment by dschult** [comment]
aha! Thanks for pointing out my error... I had the adjacency entry A_ij representing an edge from j to i instead of i to j. This is the reverse of how it is defined in NX. So when we take the out_degree we are summing each row instead of each column. That also sent me back to the paper you cited and the wording may be clarified there as well.

As you have said, the paper looks at roots that are "sinks" in that all edges from other nodes go toward the sink.  They are roots in the sense that every node has a path to the root. They emphasize this directionality by calling the graph a "sink-rooted digraph" if it has such a root. Of course, they could have written about "source-rooted digraphs" instead with similar results. The other name that they carefully qualify in this discussion is the analog of "tree" for digraphs: they discuss the number of "reverse-arborescences". The adjective "reverse" indicates that these are the arborescences of the graph with reversed edges.  I am fairly sure that the "sink-rooted digraphs" which have "reverse-arborescences" are completely analogous to "source-rooted digraphs" which have "arborescences". 

The theorem in the paper gives the number of "reverse-arborescences" for a given "sink-root". But the same method applied to the reversed digraph would give the number of "arborescences" for a given "source-root". Reversing the graph changes the Laplacian Matrix in two ways: it transposes the adjacency matrix (off-diagonal entries of L) and it changes the degree diagonal entries to be in_degree instead of out_degree. Another way to say this is that the degree values come from the column sums of A rather than the row sums.  Switching the direction of edges and the out-degree to in-degree leaves L the same. And the sink-root becomes a source-root. So, the number of reverse-arborescences in a digraph given a sink-root is the same as the number of arborescences in the reversed digraph given a the same node as a source-root.

The terms reverse-arborescence and sink-root are the opposite direction of how we use the term root and arborescence elsewhere in NX.  For BFS and DFS traversal, if we start at a root, and move in the direction of the edges, we should traverse the tree.  The BFS and DFS traversals would end/backtrack at a leaf. There could be many leafs but only one root. Thus our roots are source-roots. And discussing this function in terms of reverse-arborescences makes the wording more convoluted, especially for people coming to this from other places where arborescences are used.

Can we switch to use the term "root" and "arborescence" for this PR to mean source-root and arborescence?
I think that means we need to compute L using the in_degree (or `A.sum(axis=0)`). Everything else in the code would be the same. But we'd want to switch back to the term "root" instead of leaf. And for minimal changing of tests, we should reverse the directions of the edges in the tests.

Thoughts?     Thanks!

**Comment by peijenburg** [comment]
Thanks @dschult for sorting out the right terminology.
I completely agree with your suggestions, so I reverted everything back to `root` and changed the functionality and the tests accordingly. Hope that everything is ready now.

I tried adding the deprecation warnings, please verify whether they are correct :)

Also, please have a look at the conflicts, I've no idea why they appear every now and then.

File: networkx/linalg/laplacianmatrix.py
Comment (dschult) [inline networkx/linalg/laplacianmatrix.py:164]: Change end -> start but also add a warning of the deprecation to the doc_string with same message as in the deprecation warning.
```suggestion
    that start in the `root` node [3]_.

    .. deprecated:: 3.5

            `total_spanning_tree_weight` is deprecated and will be removed in v3.5.
            Use `nx.number_of_spanning_trees(G)` instead.

```

File: networkx/algorithms/tree/mst.py
Comment (dschult) [inline networkx/algorithms/tree/mst.py:1247]: On this last check-through I realized that the computed laplacian cofactor will be zero when the network is not connected. So really, this check (and presumably the check for weakly connected in directed case) aren't needed. They might save time in the rare case that someone mistakenly provides a disconnected graph. But they certainly use time in the normal case of a connected graph. (I say "normal" only because someone interested in spanning trees would usually have a connected graph, I think.)

What do you think about removing the check for connected and the check for weakly connected?

**Review by dschult — COMMENTED** [review]
Can you delete the lines after line 239 in laplacianmatrix.py starting with 
`## TODO remove this`?

Other items discussed below include: 
- extra check line to compute D1 and print D, D1
- add the same deprecation warning to the doc_string
- a suggestion that I realized is possible -- that we don't check for connected or weakly connected. The determinant computation should equal zero in those cases anyway. And it should be an unusual (rare?) case for someon to use this function with a disconnected graph.

Thanks!!

**Comment by dschult** [comment]
The merge conflicts occur when another PR that touches the same code this PR touches is merged.
That means `main` changes in a way that git can't figure out whether to use your changes or the other PRs changes. In this can a PR added functions for `random_spanning_tree` that appears on the same lines as this PRs changes. So I had to go through and tell git how to resolve those conflicts.  

This does mean that you will need to pull those changes down to your local branch before making further changes though. Hopefully we are close enough on this that it won't happen again on this PR. :)

File: networkx/algorithms/tree/mst.py
Comment (peijenburg) [reply networkx/algorithms/tree/mst.py:1252]: @dschult We could indeed remove them.
But the loss of speed is small (<10%) for connected graphs and the speedup for disconnected graphs is a factor 30. So I would advocate to leave the checks in place.

**Comment by peijenburg** [comment]
I applied all changes, except for removing the check of the connectedness of the graph.
It helps to speedup the computation significantly for disconnected graphs and causes only a small loss for connected graphs.

**Review by dschult — APPROVED** [review]
OK... let's leave the check for connectedness in there. 
I approve this PR...  It is ready for another set of eyes to look at before merging.

I went to change the deprecation number and found a rabbit hole in related functions in `laplacianmatrix.py`. Eventually I found the rabbit and realized why we needed to use the in-degree laplacian for our source-root and number of arborescence  case. Most of the rabbit hole was figuring out that the current laplacian matrix codes do in fact use the out-degree and that we are using the in-degree -- and that it is OK that we are using different degree values. It is strange to me that for the counting argument to work we need to use sink-roots with out-degree or we can use source-roots with in-degree. Backwards from what I expect...  

Anyway, my commit has a bunch of doc_string changes for laplacian_matrix functions. I also added a sentence about in-degree to the new mst.py function. Hopefully all is OK.

**Comment by peijenburg** [comment]
Looks OK to me. Thanks a lot for the help @dschult

**Comment by peijenburg** [comment]
@rossbar @MridulS is anyone of you available for reviewing this PR? Thanks so much!

File: networkx/algorithms/tree/mst.py
Comment (rossbar) [inline networkx/algorithms/tree/mst.py:1217]: Shouldn't the examples be updated to call `number_of_spanning_trees` instead?

File: networkx/linalg/laplacianmatrix.py
Comment (rossbar) [inline networkx/linalg/laplacianmatrix.py:255]: I would just as soon leave the remainder of the docstring for this deprecated function as-is, but I guess since the changes are already in place it's worth just pushing on without reviewing too carefully!

File: networkx/linalg/tests/test_laplacian.py
Comment (rossbar) [inline networkx/linalg/tests/test_laplacian.py:339]: If these tests are identical to the ones added for `TestNumberOfSpanningTrees` then let's not add these here - the code is already tested there.

File: networkx/linalg/laplacianmatrix.py
Comment (peijenburg) [reply networkx/linalg/laplacianmatrix.py:255]: I'll leave them as they are...

**Comment by peijenburg** [comment]
@rossbar Thanks for the comments! I've made appropriate changes.
There are quite some changes due to a conflict that I had to resolve.
However, there is still one test failing. Can you check what's happening?

**Review by rossbar — APPROVED** [review]
>  However, there is still one test failing. Can you check what's happening?

That's a recent test which was added to prevent returning numpy scalars in preparation for the numpy 2.0 release - a915359 should fix it. The current doc CI failures are unrelated.

This LGTM, let's go ahead and get it in. Thanks @peijenburg and @dschult !

**Pull request merged**
